storage/src/main/scala/coop/rchain/storage/regex/RegexPattern.scala

```diff
@@ -0,0 +1,364 @@
+        case (false, false) => CharClassPattern(this.charSet & thatCharClass.charSet)
+      }
+    case _ => multiply(1) & that
+  }
+}
+
+/**
+  * A min and a max. The vast majority of characters in regular
+  * expressions occur without a specific multiplier, which is implicitly
+  * equivalent to a min of 1 and a max of 1, but many more have explicit
+  * multipliers like "*" (min = 0, max = inf) and so on.
+  * Although it seems odd and can lead to some confusing edge cases, we do
+  * also permit a max of 0 (iff min is 0 too). This allows the multiplier
+  * "zero" to exist, which actually are quite useful in their own special way.
+  */
+final case class Multiplier(min: Option[Int], max: Option[Int]) {
+  def this(min: Int, max: Int) = this(Some(min), Some(max))
+  def this(multiplier: Int) = this(Some(multiplier), Some(multiplier))
+
+  require(
+    (min, max) match {
+      case (Some(x), Some(y)) => (0 <= x) && (x <= y)
+      case (Some(x), None)    => 0 <= x //min valid, max infinite
+      case (None, Some(_))    => false //min is infinite
+      case (None, None)       => true
+    },
+    "Invalid multiplier bounds"
+  )
+
+  /**
+    * We need some math operation on Option[Int], assuming that None == infinite
+    */
+  implicit class OptionIntMath(opt: Option[Int]) {
+    def +(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None)    => None
+      case (None, Some(_))    => None
+      case (None, None)       => None
+      case (Some(x), Some(y)) => Some(x + y)
+    }
+
+    /**
+      * Subtract another bound from this one.
+      * Caution: this operation is not meaningful for all bounds.
+      */
+    def -(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (Some(_), None) =>
+        throw new IllegalArgumentException("Can't substract infinity") //something - infinity => invalid operation
+      case (None, None)       => Some(0) //Infinity minus infinity is zero.
+      case (None, _)          => None //inf - anything => inf
+      case (Some(x), Some(y)) => Some(x - y)
+    }
+
+    def *(that: Option[Int]): Option[Int] = (opt, that) match {
+      case (None, _)          => None
+      case (_, None)          => None
+      case (Some(x), Some(y)) => Some(x * y)
+    }
+
+    def >=(that: Option[Int]): Boolean = (opt, that) match {
+      case (None, _)          => true
+      case (Some(_), None)    => false
+      case (Some(x), Some(y)) => x >= y
+    }
+
+    def <=(that: Option[Int]): Boolean = (opt, that) match {
+      case (_, None)          => true
+      case (None, Some(_))    => false
+      case (Some(x), Some(y)) => x <= y
+    }
+  }
+
+  def mandatory: Option[Int] = min
+  def optional: Option[Int] = max - min
+
+  override def toString: String = (min, max) match {
+    case (Some(_), Some(0)) =>
+      throw new ClassFormatError("Can't serialise a multiplier with max bound 0")
```

Comment:
I think it is better, in this case, to return a string with an error message, not sure.

Reply:
Fixed.